src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs

```diff
@@ -1195,5 +1197,2044 @@ static void Main()
+    public ref struct InterpolatedStringBuilder
+    {
+        public static InterpolatedStringBuilder Create(int literalLength, int formattedCount) => new InterpolatedStringBuilder(literalLength);
+        private readonly StringBuilder _builder;
+        public InterpolatedStringBuilder(int literalLength)
+        {
+            _builder = new StringBuilder();
+        }
+        public string ToStringAndClear() => _builder.ToString();
+        public void AppendLiteral(CustomStruct s) => _builder.AppendLine(s.ToString());
+        public void AppendFormatted(object o) => _builder.AppendLine(""value:"" + o.ToString());
+    }
+}";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (4,21): error CS1503: Argument 1: cannot convert from 'string' to 'CustomStruct'
+                // Console.WriteLine($"Text{1}");
+                Diagnostic(ErrorCode.ERR_BadArgType, "Text").WithArguments("1", "string", "CustomStruct").WithLocation(4, 21)
+            );
+        }
+
+        [Fact]
+        public void InvalidBuilderReturnType()
+        {
+            var source = @"
+using System;
+
+Console.WriteLine($""Text{1}"");
+
+namespace System.Runtime.CompilerServices
+{
+    using System.Text;
+    public ref struct InterpolatedStringBuilder
+    {
+        public static InterpolatedStringBuilder Create(int literalLength, int formattedCount) => new InterpolatedStringBuilder(literalLength);
+        private readonly StringBuilder _builder;
+        public InterpolatedStringBuilder(int literalLength)
+        {
+            _builder = new StringBuilder();
+        }
+        public string ToStringAndClear() => _builder.ToString();
+        public int AppendLiteral(string s) => 0;
+        public int AppendFormatted(object o) => 0;
+    }
+}";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (4,21): error CS9001: Interpolated string builder method 'InterpolatedStringBuilder.AppendLiteral(string)' is malformed. It does not return 'void' or 'bool'.
+                // Console.WriteLine($"Text{1}");
+                Diagnostic(ErrorCode.ERR_InterpolatedStringBuilderMethodReturnMalformed, "Text").WithArguments("System.Runtime.CompilerServices.InterpolatedStringBuilder.AppendLiteral(string)").WithLocation(4, 21),
+                // (4,25): error CS9001: Interpolated string builder method 'InterpolatedStringBuilder.AppendFormatted(object)' is malformed. It does not return 'void' or 'bool'.
+                // Console.WriteLine($"Text{1}");
+                Diagnostic(ErrorCode.ERR_InterpolatedStringBuilderMethodReturnMalformed, "{1}").WithArguments("System.Runtime.CompilerServices.InterpolatedStringBuilder.AppendFormatted(object)").WithLocation(4, 25)
+            );
+        }
+
+        [Fact]
+        public void MixedBuilderReturnTypes_01()
+        {
+            var source = @"
+using System;
+
+Console.WriteLine($""Text{1}"");
```

Comment:
>$""Text{1}"" [](http://example.com/codeflow?start=18&length=12)

Is it possible to test scenario when AppendLiteral is called second? For example, ```$""{1}Text""```?